src/shared/util.js

```diff
@@ -208,13 +208,15 @@ export function genStaticKeys (modules: Array<ModuleOptions>): string {
  * if they are plain objects, do they have the same shape?
  */
 export function looseEqual (a: mixed, b: mixed): boolean {
-  /* eslint-disable eqeqeq */
-  return a == b || (
-    isObject(a) && isObject(b)
-      ? JSON.stringify(a) === JSON.stringify(b)
-      : false
-  )
-  /* eslint-enable eqeqeq */
+  const isObjectA = isObject(a)
+  const isObjectB = isObject(b)
+  if (isObjectA && isObjectB) {
+    return JSON.stringify(a) === JSON.stringify(b)
+  } else if (!isObjectA && !isObjectB) {
```

Comment:
We need strict equality for primitive type, e.g., '0' is equal to false.  I'll update the test case like bellow
```js
new Vue({
  data: {
    val: false,
    test: 0
  },
  template:
    '<div>' +
      '<input type="radio" value="" v-model="test">' +
      '<input type="radio" value="0" v-model="test">' +
      '<input type="radio" :value="val" v-model="test">' +
    '</div>'
}).$mount('#app');
```